File: CS50/Week 6/readability/readability.py
import re


def main() :
    text = get_string("Text: ")
    nr_letters = get_letters(text)
    nr_words = get_words(text)
    nr_sentences = get_sentences(text)
    print_grade(nr_letters, nr_words, nr_sentences)
    return 0


def get_string(string) :
    return input(string)


def get_letters(text) :
    letters_dict = dict()
    for letter in text :
        if letter.isalpha() == True : letters_dict[letter] = letters_dict.get(letter, 0) + 1
    nr_letters = 0
    for letter, num in letters_dict.items() :
        nr_letters += num
    return nr_letters


def get_words(text) :
    return len(re.findall('[\S]+', text))


def get_sentences(text) :
    return len(re.findall('\S[^.!?]*[.!?]', text))


def print_grade(nr_letters, nr_words, nr_sentences) :
    L = 100 * nr_letters / nr_words
    S = 100 * nr_sentences / nr_words
    grade = 0.0588 * L - 0.296 * S - 15.8
    grade = int(grade) if grade - int(grade) < 0.5 else int(grade) + 1
    if grade <= 1 :
        print("Before Grade 1")
    elif grade > 1 and grade < 16 :
        print(f"Grade {grade}")
    elif grade >= 16 :
        print("Grade 16+")


main()

File: CS50/Week 2/Problem Set 2/readability/readability.c
#include <ctype.h>
#include <cs50.h>
#include <stdio.h>
#include <string.h>

int main(void)
{
	// ask the user for input
	
	string passage = get_string("Text: ");

	int nr_letters = 0;
	int nr_words = 0;
	int nr_sentences = 0;

	// compute nr of letters
	// compute nr of words
	// compute nr of sentences
	
	int passage_length = strlen(passage);

	for (int i = 0; i < passage_length; i++)
	{
		// if sentence increment
		if ((passage[i + 1] == '.' || passage[i + 1] == '?' || passage[i+1] == '!' || passage[i+1] == '\0') && (isalnum(passage[i]) || passage[i] == ' '))
		{
			nr_sentences++;
		}
		// if letter increment
		if (isalpha(passage[i]))
		{
			nr_letters++;
		}
		// if word increment
		if (passage[i] == ' ')
		{
			nr_words++;
		}
	}

	// count the last word which doesn't have a space after it
	nr_words++;

	// compute nr of average letters per 100 words
	// compute nr of average sentences per 100 words
	
	float L = (float) nr_letters * 100.0 / (float) nr_words;
	float S = (float) nr_sentences * 100.0 / (float) nr_words;

	// compute the index and print it
	float index = 0.0588 * L - 0.296 * S - 15.8;

	// round the index to its nearest whole number
	if (index - (int) index < 0.5)
	{
		index = (int) index;
	}
	else
	{
		index = (int) index + 1;
	}

	if (index < 1)
	{
		printf("Before Grade 1\n");
	}
	else if (index >= 16)
	{
		printf("Grade 16+\n");
	}
	else
	{
		printf("Grade %i\n", (int) index);
	}

	return 0;
}
